---
title: "Complex_diagnostics_paper"
author: "Swetansu Pattnaik"
date: "17/01/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#.libPaths(c( "/home/shu/R/x86_64-redhat-linux-gnu-library/3.4", .libPaths() ) )
`%nin%` = Negate(`%in%`)
library(SKAT)
library(ggplot2)
library(cowplot)
```


##functions
#prepare SKAT output
```{r}
get_gene_pval <- function(skatlist_inp)
{
  skat_df <- do.call("rbind.data.frame", skatlist_inp)
  skat_df_sub =  skat_df[skat_df[,1] %in% all_genes,]
  obj <- deparse(substitute(skatlist_inp))
  skat_df_sub$Class = obj
  cnames = colnames(skat_df_sub)
  cnames = gsub("pval_SKATbin", "pval_SKATburden", cnames)
  colnames(skat_df_sub) = cnames
  return(skat_df_sub)
}

```


##Data
```{r}
##ISKS vs MGRB
isks_mgrb_genes <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/Exome_para_tab_ISKS_MGRB_C345_Aug31.rds")
isks_mgrb_genes_noC3 <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/Exome_para_tab_ISKS_MGRB_minusC3_Aug31.rds")

##ISKS vs MGRB (minus_71)
isks_mgrb_genes_cc <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/Exome_para_tab_ISKS_MGRB_C345_Jan10_m71_ccrect.rds")
isks_mgrb_genes_noC3_cc <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/Exome_para_tab_ISKS_MGRB_minusC3_Jan10_m71_ccrect.rds")

```

##SKAT results from different mixed models
```{r}

std_PC1234 = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/Exome_skat_para_result_isks_combset2020_uni_MAF_PC1234_ver4_clinrect_Aug31.rds")
std_PCs = std_PC1234[[4]]
#std_PC1234_min13 = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/SKAT/Exome_skat_para_result_isks_totex_count_PC1234_ver4_clinrect_Aug31_subPC1.rds")
#std_PCs_min13 = std_PC1234_min13[[4]]
std_PC1234_min71_cc = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/SKAT/Exome_skat_para_result_isks_totex_count_PC1234_ver4_clinrect_Jan10_ccrect_PCnew.rds")
std_PC1234_EMMAX_min71_cc = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/SKAT/Exome_skat_para_result_isks_totex_count_PC1234_subPC1_EMMAX_Jan10_ccrect_PCnew.rds")
std_grm_GCTA_min71_cc = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/SKAT/Exome_skat_para_result_isks_totex_count_noPC1234_ver4_clinrect_grmGCTA_Jan10_ccrect_PCnew.rds")
#std_PCs_EUR = readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/SKAT/Exome_skat_para_result_isks_totex_count_PC1234_ver4_clinrect_Jan10_EUR.rds")


```

##Gene complexes
```{r}
##for paper : Nov-19-2021
Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")

CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")
#MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")
##for paper : Nov-19-2021
MPNST_pos <- c("NF1")

all_genes = unique(c(Shelterin, CEP_HAUS_core, MPNST_pos, "TP53", "BRCA2", "BRCA1", "PALB2"))

```


##Extract SKAT burden test pvalues from genes from complexes and report as a dataframe

```{r}
std_PC1234_cpx = get_gene_pval(std_PCs)
std_PC1234_cpx = std_PC1234_cpx[,c(1,3,6)]
colnames(std_PC1234_cpx)[2] = "pval_SKATburden"
std_PC1234_min71_cc_cpx = get_gene_pval(std_PC1234_min71_cc)
std_PC1234_EMMAX_min71_cc_cpx = get_gene_pval(std_PC1234_EMMAX_min71_cc)
std_grm_GCTA_min71_cc_cpx = get_gene_pval(std_grm_GCTA_min71_cc)
#std_PC1234_EUR_cpx = get_gene_pval(std_PCs_EUR)

```

##Prepare combined plot for all models

```{r}
##for fig S7
comb_df_pval = rbind.data.frame(std_PC1234_cpx, std_PC1234_min71_cc_cpx, std_PC1234_EMMAX_min71_cc_cpx, std_grm_GCTA_min71_cc_cpx)

comb_df_pval$log10pval = -log10(comb_df_pval$pval_SKATburden)

samp_ord = c("TP53", MPNST_pos, Shelterin, CEP_HAUS_core,  "BRCA1", "BRCA2", "PALB2")
comb_df_pval$eg_ID <- factor(comb_df_pval$eg_ID,levels=samp_ord,ordered=TRUE)

##rename model names

comb_df_pval$Class = gsub("std_PCs", "PCs_base",comb_df_pval$Class)
comb_df_pval$Class = gsub("std_PC1234_min71_cc", "PCs_min71",comb_df_pval$Class)
comb_df_pval$Class = gsub("std_PC1234_EMMAX_min71_cc", "PCs_nullEMMAX_min71",comb_df_pval$Class)
comb_df_pval$Class = gsub("std_grm_GCTA_min71_cc", "stdGRM_nullEMMAX_min71",comb_df_pval$Class)

##rename complex names: for paper Nov-19-2021
comb_df_pval$complex = ifelse(comb_df_pval$eg_ID %in% Shelterin, "Shelterin", 
                              ifelse(comb_df_pval$eg_ID %in% CEP_HAUS_core, "Centrosome", 
                                     ifelse(comb_df_pval$eg_ID %in% MPNST_pos , "NF1",
                                            ifelse(comb_df_pval$eg_ID %in% c("BRCA1", "BRCA2", "PALB2"), "HBOC", "TP53"))))


```

##plots
```{r}
###Final plot
p1 = ggplot(comb_df_pval, aes(x=eg_ID, y=log10pval)) +  
  geom_boxplot(show.legend = F) +
  geom_point(position=position_jitterdodge(jitter.width=0, dodge.width = 0.3), 
             aes(color=factor(Class)), show.legend = T) +
  geom_hline(yintercept = 1, lty=2, colour = "black") +
  theme_cowplot(font_size = 13) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "bottom", legend.title = element_blank()) +
  xlab("") + ylab("-log10(P.value)") 

##option 2
p2 = ggplot(comb_df_pval, aes(x=eg_ID, y=log10pval)) +  
  geom_boxplot(show.legend = F) +
  geom_point(position=position_jitterdodge(jitter.width=0, dodge.width = 0.3), 
             aes(color=factor(Class)), show.legend = T) +
  geom_hline(yintercept = 1, lty=2, colour = "black") +
  theme_cowplot(font_size = 12) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "top", legend.title = element_blank()) +
  xlab("") + ylab("-log10(P.value)") 

```


##Fisher's test

##function to make input table for Fisher test
```{r}

##make table for hypergeometric test
tab_fisher = function(comb_model_cpx_genes, count_tab){
  classes = unique(comb_model_cpx_genes$Class)
  if(length(classes) > 1) {
  sig_genes_all = list()
  for(i in 1:length(classes)){
  sig_genes = comb_model_cpx_genes[comb_model_cpx_genes$Class %in% classes[i] & 
                                     comb_model_cpx_genes$pval_SKATburden < 0.1, ]
  sig_genes[,6:8] = count_tab[match(sig_genes[,1], count_tab$gene),1:3 ]
  colnames(sig_genes[6:8]) = c("case", "control", "gene")
  sig_genes_all[[i]] = sig_genes
  }
  sig_genes_all_df = do.call("rbind.data.frame", sig_genes_all)
  return(sig_genes_all_df)
  }
  else{
    sig_genes = comb_model_cpx_genes[comb_model_cpx_genes$Class %in% classes & 
                                       comb_model_cpx_genes$pval_SKATburden < 0.1, ]
    print(dim(sig_genes))
    sig_genes[,4:6] = count_tab[match(sig_genes[,1], count_tab$gene),1:3 ]
    colnames(sig_genes[4:6]) = c("case", "control", "gene")
    return(sig_genes)
  }
  #  sig_genes = sig_genes[!is.na(sig_genes$ISKS),]
}


```


##data for fisher test

```{r}
##make separate combined_df_pval dataframe for different cohort sizes
##with case control optimization
comb_df_pval_counts_noC3_cc = tab_fisher(comb_df_pval, isks_mgrb_genes_noC3_cc) 
# PC1234_min71_pval_counts_noC3_m71 = tab_fisher(std_PC1234_min71_cc_cpx, isks_mgrb_genes_noC3_cc) 
# GRM_min71_pval_counts_noC3_m71 = tab_fisher(std_grm_GCTA_min71_cc_cpx, isks_mgrb_genes_noC3_cc) 
# comb_df_pval_counts_noC3_m71 = rbind.data.frame(PC1234_min71_pval_counts_noC3_m71, GRM_min71_pval_counts_noC3_m71)
# comb_df_pval_counts_noC3_m71$complex = comb_df_pval_counts_noC3_cc[match(comb_df_pval_counts_noC3_m71$eg_ID, comb_df_pval_counts_noC3_cc$eg_ID),5]
# 
# 
##original data
comb_df_pval_counts_noC3 = tab_fisher(std_PC1234_cpx, isks_mgrb_genes_noC3)

comb_df_pval_counts_noC3$complex = comb_df_pval_counts_noC3_cc[match(comb_df_pval_counts_noC3$eg_ID, comb_df_pval_counts_noC3_cc$eg_ID),5]
comb_df_pval_counts_noC3 = comb_df_pval_counts_noC3[,c(1:3,6:7,4:5)]

```


##function for fisher test; table S13 generated in this step
```{r}
##Run Fisher test for all models

cpx_OR_fisher_count <- function(comb_df_inp, mod_name, case_coh_size, cont_coh_size, cpx_name, sub_case, sub_cont, sub_ISKS, sub_MGRB){
  print(cpx_name)
  print(mod_name)
  comb_df_inp_tab <- comb_df_inp[comb_df_inp$complex %in% cpx_name & comb_df_inp$Class %in% mod_name,]
  print((comb_df_inp_tab))
  # ppi_res_tab[,2] <- ifelse(ppi_res_tab[,2] == 0, 1, ppi_res_tab[,2])
  # inp <- c(sum(ppi_res_tab[,1]) - sub_case , case_coh_size - sum(ppi_res_tab[,1]) - sub_ISKS , 
  #          sum(ppi_res_tab[,2]) - sub_cont, cont_coh_size - sum(ppi_res_tab[,2]) - sub_MGRB)
  inp <- c(sum(comb_df_inp_tab[,6], na.rm = T) - sub_case , case_coh_size - sub_ISKS - (sum(comb_df_inp_tab[,6], na.rm = T) - sub_case), 
           sum(comb_df_inp_tab[,7],na.rm = T) - sub_cont, cont_coh_size - sub_MGRB - (sum(comb_df_inp_tab[,7], na.rm = T) - sub_cont))
  sim_mat <- matrix(inp ,nrow = 2, ncol = 2)
  colnames(sim_mat) <- c("case", "cont")
  rownames(sim_mat) <- c("hits", "no_hits")
  print(sim_mat)
  #ft <- fisher.test(sim_mat, alternative = "greater")
  #ft <- fisher.test(sim_mat)
  #cpx_lab = deparse(substitute(cpx_name))
  ft <- fisher.test(sim_mat, conf.int = T, conf.level = 0.95)
  ft_df <- cbind.data.frame("gene" = cpx_name ,"Cases" = sim_mat[1,1],
                            "Controls" = sim_mat[1,2],
                            "Fish_pval" = ft$p.value,"CI_lower" = ft$conf.int[1],
                            "CI_upper" = ft$conf.int[2],
                            "OR_Fish" = ft$estimate, "case_coh_size" = sum(sim_mat[,1]),
                            "cont_coh_size" = sum(sim_mat[,2]), "model" = mod_name)
  return(ft_df)
}

##reports only genes that are significant in SKATburden test and are present in the complex
##CC; minus 71
cpx_res_list_noC3 = list()
cpx_res_list_noC3[[1]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "PCs_min71", 1575, 3203, "Shelterin", 0, 0, 0, 0)
cpx_res_list_noC3[[2]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "PCs_min71", 1575, 3203, "Centrosome", 0, 0, 0, 0)
cpx_res_list_noC3[[3]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "PCs_nullEMMAX_min71", 1575, 3203, "Shelterin", 0, 0, 0, 0)
cpx_res_list_noC3[[4]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "PCs_nullEMMAX_min71", 1575, 3203, "Centrosome", 0, 0, 0, 0)
cpx_res_list_noC3[[5]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "stdGRM_nullEMMAX_min71", 1575, 3203, "Shelterin", 0, 0, 0, 0)
cpx_res_list_noC3[[6]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3_cc, "stdGRM_nullEMMAX_min71", 1575, 3203, "Centrosome", 0, 0, 0, 0)

##base model

comb_df_pval_counts_noC3 =  comb_df_pval_counts_noC3[!is.na(comb_df_pval_counts_noC3$ISKS),]
cpx_res_list_noC3[[7]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3, "std_PCs", 1644, 3205, "Shelterin", 0, 0, 0, 0)
cpx_res_list_noC3[[8]] = cpx_OR_fisher_count(comb_df_pval_counts_noC3, "std_PCs", 1644, 3205, "Centrosome", 0, 0, 0, 0)


cpx_res_list_noC3_df = do.call("rbind.data.frame", cpx_res_list_noC3)

write.table(cpx_res_list_noC3_df, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/git_proj_rvas/Comb_set_round2/ISKS_MGRB_2020/ISKS_AR_AD/Aug31_freeze/review_2/Results/Fisher_Tab_S13.tsv", sep = "\t", row.names = F, quote = F)

```


##SKAT burden test at complex level

##SKAT function
```{r}
##########################################

##SKAT null function with customised covariate
SKAT_fun_null <- function(x=NULL, p_vec_pop){
  if(is.null(x)){
    obj_N <- SKAT::SKAT_Null_Model(p_vec_pop ~ 1,out_type="D")
  }
  # else if(is.integer(x) || is.numeric()){
  else if(is.null(dim(x))){
    obj_N <- SKAT::SKAT_Null_Model(p_vec_pop ~ x,out_type="D")
  }
  else if(dim(x)[2] > 1){
    nul_for <- as.formula(paste("p_vec_pop", paste(colnames(x), collapse = " + "), sep = " ~ "))
   # obj_N <- SKAT::SKAT_Null_Model(nul_for, data = p_Data_noCH_sub, out_type="D")
      obj_N <- SKAT::SKAT_Null_Model(nul_for, data = x, out_type="D")
  }
  return(obj_N)
}

##SKAT function for SKAT, SKATBinary, SKATO, SKAT_ERA
SKAT_run <- function(geno_mat, gene, x=NULL, p_vec_pop, cust_weight = NULL, rho = NULL){
  
  if(dim(geno_mat)[2] > 1 ){
    null_d <- SKAT_fun_null(x, p_vec_pop)
    # pval_SKAT <- SKAT::SKAT(t(geno_mat), null_d, weights = cust_weight, r.corr = rho)$p.value
    pval_SKATbin <- SKAT::SKATBinary(t(geno_mat), null_d, method = "Burden", weights = cust_weight)$p.value
    skat_pv <- cbind.data.frame("eg_ID" = gene, pval_SKATbin)
    return(skat_pv)
  }
  else if(dim(geno_mat)[2] == 1){
    null_d <- SKAT::SKAT_fun_null(x, p_vec_pop)
    pval_SKATbin <- SKAT::SKATBinary(t(geno_mat), null_d, method = "Burden", weights = cust_weight)$p.value
    skat_pv <- cbind.data.frame("eg_ID" = gene, pval_SKATbin)
    return(skat_pv)
  }
}

##for complexes after removal of POPMAX variant > 0.001, with new PCs
##parallel process complexes
para_SKAT_cpx_var <- function(genes, cpx_name, rm_var_file){
  skat_pvals_sex_pc12 = list()
  
  ftemp_tab <- fil_tab[fil_tab$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  ftemp_tab <- ftemp_tab[ftemp_tab$VARIANT %nin% rm_var_file,]
  
  print(max(ftemp_tab$comb_score))
  if(dim(ftemp_tab) == 0 ){
    next
  }
  else{
    ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
    samp_vec <- list()
    for(m in 1:length(ftemp_tab_var_id)){
      sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
      sam_gene_gt <- unique(sam_gene_gt)
      if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
        # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
        vep_con <- unique(sam_gene_gt$vep_consequence)
        samp_vec_con <- list()
        for(k in 1:length(vep_con)){
          sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
          sam_gene_gt_con <- unique(sam_gene_gt_con)
          maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          ##genotype matrix  
          sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
          sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt_con$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
          samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                 unique(sam_gene_gt_con$gene_symbol), 
                                 unique(sam_gene_gt_con$vep_consequence), 
                                 unique(as.character(sam_gene_gt_con$auto_call)),
                                 as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                 as.numeric(maf_vec_cont), sam10)
        }
        #  vep_cod <- sapply(str_extract_all(sam_gene_gt$vep_Codons, "[A-Z]+"), paste, collapse= '_')
        #  ref_alt <- unlist(lapply(strsplit(sam_gene_gt$VARIANT, split = ":"), function(x) paste(x[3], x[4], sep = "_")))
        #  sam_gene_gt <- sam_gene_gt[which(vep_cod %in% ref_alt),]
        
      }
      else{
        ##compute cohort specific MAF
        maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                     grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
        ##genotype matrix  
        sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
        sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt$SAMPLE, 1, 0)
        sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
        ##account for discrepancy in vep_consequence
        #vep_con <- names(sort(table(unlist(strsplit(sam_gene_gt$vep_consequence, split = "&"))),decreasing=TRUE)[1])
        # sam_gene_gt$vep_consequence <- vep_con
        
        
        samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                           unique(sam_gene_gt$gene_symbol), 
                           unique(sam_gene_gt$vep_consequence), 
                           unique(as.character(sam_gene_gt$auto_call)),
                           as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                           as.numeric(maf_vec_cont), sam10)
      }
    }
    samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
    colnames(samp_vec_mat_uni) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                    "coh_MAF_cont", Ex_samp_id)
    if(exists("samp_vec_con")){
      samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
      colnames(samp_vec_mat_con) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id)
      samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
    }else{
      samp_vec_mat <- samp_vec_mat_uni
    }
    print(dim(samp_vec_mat))
    ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
    ##changed to uniform intra_cohort_MAF filter
    ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
    #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
    #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) >= 0.00035 | 
                                     as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) >= 0.00035),]
    #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    
    if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
      
      skat_pvals_sex_pc12 <- NULL
      
      next
    }
    ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
    
    else {
      gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
      gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
      class(gene_mat) <- "numeric"
      
      ##gender + totex + pc1:4
      skat_pvals_sex_pc12 <- SKAT_run(geno_mat = gene_mat, gene = cpx_name, x=p_Data_noCH[,c(54:55,19:22)], p_vec, cust_weight = gene_mat_comb, rho = 1)
      
    }
    
  } ##end of nested else loop
  return(skat_pvals_sex_pc12)
}##end of gene loop ; i loop   

```


##SKAT input
```{r}
##use fil_tab and PCs after removal of samples using centroid method
fil_tab <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/all_isksmgrb_skatinp_combset2020_clin_C3C4C5_NFE0002_AD_rm_dup_freeze_rev_PC1_sub.tsv", sep = "\t", header = T, stringsAsFactors = F) ##generated by R2Q6 script pca_pop*ISKS.R
fil_tab <- fil_tab[!is.na(fil_tab$SAMPLE),]
dim(fil_tab)

##Additive model
Ex_samp_id <- unique(fil_tab$SAMPLE)

######get phenotype data to control for age and sex and PC's; 

QC2_dat <- read.table("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/Rep_set_variants/joint_calls_2019jan_2019nov.final_qc_output.tsv", 
                      header = T, sep = "\t", stringsAsFactors = F)
QC2_dat_pass <- QC2_dat[QC2_dat$passes_qc2 %in% "TRUE",]
QC2_dat_pass$isFemale <- ifelse(QC2_dat_pass$f_stat < 0.2, 1, 
                                ifelse(QC2_dat_pass$f_stat > 0.8, 0, 2))
##for future QC with prob.NFE correction use the following file and adjust the column for gender from 39 to 53

##p_Data_noCH below is generated by pca_tsne_diagnostics.Rmd script: Added country and cluster
##p_Data_noCH <- readRDS("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_PCnew.rds")
p_Data_noCH <- read.delim("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_rev2_combset_pca.scores_clustered.tsv", sep = "\t", header = T, stringsAsFactors = F) 
p_Data_noCH$rect_sam = p_Data_noCH$sample # the samples name are rectified at HAIL stage
p_Data_noCH <- p_Data_noCH[as.character(p_Data_noCH$rect_sam) %in% Ex_samp_id,]
##drop MGRB sample BAAUD; not in latest call
#samp_ID_match <- samp_ID_match[grep("BAAUD", samp_ID_match, invert = T)]
#p_Data_noCH <- p_Data_noCH[match(samp_ID_match, p_Data_noCH$sample),]
Ex_samp_id <- Ex_samp_id[match(p_Data_noCH$rect_sam, Ex_samp_id)]
##filter out QC fail cases
fil_tab <- fil_tab[fil_tab$SAMPLE %in% Ex_samp_id,]

p_Data_noCH$gender <- QC2_dat_pass[match(p_Data_noCH$rect_sam, QC2_dat_pass$new_sampleid), 21]
#write.table(p_Data_noCH, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/PID/pheno_data_SKAT_final_freeze.tsv",
#            sep = "\t", row.names = F, quote = F)

##Add exome count
##ref: Exome sequencing in amyotrophic lateral sclerosis implicates a novel gene, DNAJC7, encoding a heat-shock protein
##total exome count (summation of synonymous variants, benign missense variants, damaging missense variants and PTVs) 
tot_exome_count <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/all_isks_mgrb_total_exome_count_clingene_rnd3_freeze.tsv",
                              sep = "\t", header = T, stringsAsFactors = F)
p_Data_noCH$totex_count <- tot_exome_count[match(p_Data_noCH$rect_sam, tot_exome_count$Var1), 2]

#binary phenotype vector 
p_vec <- ifelse(!is.na(as.numeric(as.character(p_Data_noCH$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH$rect_sam)), 1, 0)

```


##SKAT burden test run after collapsing variants across all the genes in a complex.
##Note that the variants were filtered using POPMAX_AF < 0.001.
##POPMAX scores were derived from dbnsfp.


```{r echo=FALSE, message=FALSE}

Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")

Sarcoma_genes <- c("TP53", "NF1", "SDHA", "SDHB", "SDHD")
CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")

Breast_cancer_genes = c("BRCA2", "BRCA1", "PALB2")

MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")

cpx_list <- list(Shelterin, CEP_HAUS_core, Breast_cancer_genes, Sarcoma_genes, MPNST_pos)
names(cpx_list) <- c("Shelterin", "Centrosome", "BRCA_genes", "SARC_genes", "MPNST")

##Parallelised SKAT  
library(doParallel)
library(doMC)
registerDoMC(30)

##POPMAX filtered variants
dbnsfp_cpx = read.delim("~/RVAS/dbNSFP_4.1a/isks_cpx_var_dbnsfp_out_fin.tsv", sep = "\t", header = T, stringsAsFactors = F)
for(i in 6:9){
  dbnsfp_cpx[,i] <- as.numeric(dbnsfp_cpx[,i])
}
dbnsfp_cpx[is.na(dbnsfp_cpx)] = 0
dbnsfp_cpx =  unique(dbnsfp_cpx)
rownames(dbnsfp_cpx) = gsub(" ", "", apply(dbnsfp_cpx[ ,1:4] , 1 , paste, collapse = ":" ))

dbnsfp_cpx_ind_01 = apply(dbnsfp_cpx[,6:9], 1, function(x)ifelse(max(x) >= 0.001, 1, 0))
dbnsfp_cpx_samp_01 = names(dbnsfp_cpx_ind_01[dbnsfp_cpx_ind_01 == 1])


res20_cpx_var_pc <- list()
#genes <- unique(fil_tab$gene_symbol)
#genes <- names(cpx_list)
system.time(res20_cpx_var_pc <- foreach(i=1:length(cpx_list), .errorhandling = 'remove') %dopar% 
{para_SKAT_cpx_var(cpx_list[[i]], names(cpx_list)[i], dbnsfp_cpx_samp_01)}) ##from POPMAX : line 454

res20_cpx_var_pc_df = do.call("rbind.data.frame", res20_cpx_var_pc)
res20_cpx_var_pc_df$p.adj = p.adjust(res20_cpx_var_pc_df$pval_SKATbin)
write.table(res20_cpx_var_pc_df, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/git_proj_rvas/Comb_set_round2/ISKS_MGRB_2020/ISKS_AR_AD/Aug31_freeze/review_2/Results/SKAT_complex_level_output.tsv", sep = "\t", row.names = F, quote = F)

```

