---
title: "Population stratification rev 2"
author: "Swetansu Pattnaik"
date: "08/01/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##libraries
```{r echo=FALSE, message=FALSE}
library(ggplot2)
library(readxl)
library(mclust)
library(gg3D)
library(dplyr)
library(kableExtra)
`%nin%` = Negate(`%in%`)
```

##Functions
#3D plots
```{r echo=FALSE, message=FALSE}
source("~/APCluster_graphs/case_TCGA_new_PRAD_3431/multi_plot.R")
library("gg3D")

make_3Dplot_clust <- function(df_inp, theta=0, phi=0){
  ggplot(df_inp, aes(x=PC1, y=PC2, z=PC3, colour=clust_code)) +
    axes_3D(theta=theta, phi=phi) +
    stat_3D(theta=theta, phi=phi, geom="point") +
    labs_3D(theta=theta, phi=phi, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 
}

#clust3D_12 = make_3Dplot_clust(scores_sub_PC1_mem, theta=-90, phi=60)

make_3Dplot_coh <- function(df_inp, theta=0, phi=0){
  ggplot(df_inp, aes(x=PC1, y=PC2, z=PC3, colour=superPopulation)) +
    axes_3D(theta=theta, phi=phi) +
    stat_3D(theta=theta, phi=phi, geom="point") +
    labs_3D(theta=theta, phi=phi, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 
}

#coh3D_12 = make_3Dplot_coh(scores_sub_PC1_mem, theta=-90, phi=60)

##For All populations
make_3Dplot_pop <- function(df_inp, theta=0, phi=0){
  ggplot(df_inp, aes(x=PC1, y=PC2, z=PC3, colour=pred.superPop)) +
    axes_3D(theta=theta, phi=phi) +
    stat_3D(theta=theta, phi=phi, geom="point") +
    labs_3D(theta=theta, phi=phi, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 
}


##For European population

make_3Dplot_eurpop <- function(df_inp, theta=0, phi=0){
  ggplot(df_inp, aes(x=PC1, y=PC2, z=PC3, colour=pred.eurPop)) +
    axes_3D(theta=theta, phi=phi) +
    stat_3D(theta=theta, phi=phi, geom="point") +
    labs_3D(theta=theta, phi=phi, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 
}

#eurpop3D_12 = make_3Dplot_eurpop(scores_EUR, theta=180, phi=0)


```


##2D plots

```{r echo=FALSE, message=FALSE}

#for case control
plot_case_cont_2D = function(df_inp){
plot12_rect_cc <- ggplot(df_inp, aes(x = PC1, y = PC2, colour = superPopulation)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 
plot12_rect_pop <- ggplot(df_inp, aes(x = PC1, y = PC2, colour = pred.superPop)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 
plot12_rect_country <- ggplot(df_inp[df_inp$Country %nin% "Unknown",], aes(x = PC1, y = PC2, colour = Country)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 

multiplot(plot12_rect_cc, plot12_rect_pop, plot12_rect_country, cols = 3)

}

plot_eur_case_cont_2D = function(df_inp){
plot12_rect_cc <- ggplot(df_inp, aes(x = PC1, y = PC2, colour = superPopulation)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 
plot12_rect_eurpop <- ggplot(df_inp, aes(x = PC1, y = PC2, colour = pred.eurPop)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 
plot12_rect_country <- ggplot(df_inp[df_inp$Country %nin% "Unknown",], aes(x = PC1, y = PC2, colour = Country)) + geom_point() + theme_bw() + theme(legend.position="bottom", legend.title = element_blank()) 

multiplot(plot12_rect_cc, plot12_rect_eurpop, plot12_rect_country, cols = 3)

}

#for clustered output
plot_clust_2D = function(df_inp){
clust_plot12 <- ggplot(df_inp, aes(x = PC1, y = PC2, colour = clust_code)) + geom_point() + theme_bw() + 
  theme(legend.position="bottom", legend.title = element_blank()) + 
  geom_vline(xintercept = -0.2, lty = 2)
clust_plot13 <- ggplot(df_inp, aes(x = PC1, y = PC3, colour = clust_code)) + geom_point() + theme_bw() + 
  theme(legend.position="bottom", legend.title = element_blank()) + 
  geom_vline(xintercept = -0.2, lty = 2)
clust_plot23 <- ggplot(df_inp, aes(x = PC2, y = PC3, colour = clust_code)) + geom_point() + theme_bw() + 
  theme(legend.position="bottom", legend.title = element_blank()) + 
  geom_vline(xintercept = -0.2, lty = 2)

multiplot(clust_plot12, clust_plot13, clust_plot23, cols = 3)

}

```


##Case control proportion

```{r echo=FALSE, message=FALSE}

##agglomorative clustered data
case_cont_ratio_clust = function(df_inp, pop){
case_cont_df = list()
for(i in 1:length(unique(df_inp$cluster))){
  #print(i)
 
  clust_coh_prop = as.data.frame(table(df_inp[df_inp$cluster == i,]$superPopulation))
  cont = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "controls", ]$Freq)
  case = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "sarcoma", ]$Freq)
  clust_code = paste("C", i, sep = "_")
  #print(case)
  #print(cont)
  if( pop %in% "All"){
    clust_prop = as.data.frame(table(df_inp[df_inp$cluster == i,]$pred.superPop))
 pop_split = paste(paste(clust_prop$Var1, clust_prop$Freq, sep = ":"), collapse = ",")
  ratio = (case/1631)/(cont/3205)
  }
  else {
    clust_prop = as.data.frame(table(df_inp[df_inp$cluster == i,]$pred.eurPop))
 pop_split = paste(paste(clust_prop$Var1, clust_prop$Freq, sep = ":"), collapse = ",")
  ratio = (case/1514)/(cont/3116)
  }
  
  case_cont_df[[i]] = cbind.data.frame("case" = case, "control" =  cont, "ratio" = ratio, 
                                      "split" = pop_split,  "cluster" = clust_code)
}

case_cont_df_fin = do.call("rbind.data.frame", case_cont_df)
return(case_cont_df_fin)
}

##All 1000 genome superpopulations
case_cont_ratio_pop = function(df_inp){
allpop = unique(df_inp$pred.superPop) ##based on 4 PCs
case_cont_df = list()
for(i in 1:length(unique(allpop))){
  #print(i)
  #clust_prop = as.data.frame(table(scores_EUR[scores_EUR$pred.eurPop %in% eurpop[i],]$pred.superPop))
  clust_coh_prop = as.data.frame(table(df_inp[df_inp$pred.superPop %in% allpop[i],]$superPopulation))
  cont = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "controls", ]$Freq)
  case = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "sarcoma", ]$Freq)
  
  #print(case)
  #print(cont)
  ratio = (case/1631)/(cont/3205)
  #print(ratio)
  
  case_cont_df[[i]] = cbind.data.frame("case" = case, "control" =  cont, "ratio" = ratio, "pop" = allpop[i])
}
case_cont_df_fin = do.call("rbind.data.frame", case_cont_df)
return(case_cont_df_fin)
}


##European population
case_cont_ratio_eur = function(df_inp){
eurpop = unique(df_inp$pred.eurPop) ##based on 4 PCs
case_cont_eur_df = list()
for(i in 1:length(unique(eurpop))){
  #print(i)
  #clust_prop = as.data.frame(table(scores_EUR[scores_EUR$pred.eurPop %in% eurpop[i],]$pred.superPop))
  clust_coh_prop = as.data.frame(table(df_inp[df_inp$pred.eurPop %in% eurpop[i],]$superPopulation))
  cont = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "controls", ]$Freq)
  case = sum(clust_coh_prop[clust_coh_prop$Var1 %in% "sarcoma", ]$Freq)
  
  #print(case)
  #print(cont)
  ratio = (case/1514)/(cont/3116)
  #print(ratio)
  
  case_cont_eur_df[[i]] = cbind.data.frame("case" = case, "control" =  cont, "ratio" = ratio, "eurpop" = eurpop[i])
}
case_cont_eur_df_fin = do.call("rbind.data.frame", case_cont_eur_df)
return(case_cont_eur_df_fin)
}



```


##Agglomerative clustering (method = centroid): with 4 PCs

```{r echo=FALSE, message=FALSE}

perf_clust = function(df_inp, nclust, fname){
  set.seed(231735)
  
df_inp_hc = hclust(dist(df_inp[,19:22],method = "euclidean", diag = T), method = "centroid")
#plot(scores_sub_PC1_hc)

memb <- cutree(df_inp_hc, k = nclust) ## as there are 5 superpopulations

##bootstrap check
library(fpc)   
# set the desired number of clusters                               
kbest.p = nclust       

#   Run clusterboot() with hclust 
#   ('clustermethod=hclustCBI') using Ward's method 
#   ('method="ward"') and kbest.p clusters 
#   ('k=kbest.p'). Return the results in an object 
#   called cboot.hclust.
set.seed(231735)
cboot.hclust <- clusterboot(dist(df_inp[,19:22], method = "euclidean", diag = T),clustermethod=disthclustCBI, method="centroid", k=kbest.p, B = 100)
saveRDS(cboot.hclust, paste("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/cboot.hclustB100_", fname, ".rds",   sep = ""), compress = T)
#   The results of the clustering are in 
#   cboot.hclust$result. The output of the hclust() 
#   function is in cboot.hclust$result$result. 
#
#   cboot.hclust$result$partition returns a 
#   vector of clusterlabels. 
#groups <- cboot.hclust$result$partition  

# Values close to 1 indicate stable clusters
cboot.hclust$bootmean 

df_inp_mem = df_inp %>%
  mutate(cluster = memb)
return(df_inp_mem)
}


```


##tsne
```{r echo=FALSE, message=FALSE}
library(Rtsne)
library(dplyr)
library(magrittr)
library(ggrepel)

run_tsne_viz = function(df_inp){
  set.seed(678456)
tsne.norm = Rtsne(df_inp[,19:22], pca = FALSE)
info.norm = tibble(eurpop = factor(df_inp$pred.eurPop), cohort = factor(df_inp$superPopulation))
info.norm %<>% mutate(tsne1 = tsne.norm$Y[, 1], tsne2 = tsne.norm$Y[,2])
return(info.norm)
}

run_tsne_clust = function(df_inp){
  set.seed(678456)
tsne.norm = Rtsne(df_inp[,19:22], pca = FALSE)
info.norm = tibble(eurpop = factor(df_inp$pred.eurPop), cohort = factor(df_inp$clust_code))
info.norm %<>% mutate(tsne1 = tsne.norm$Y[, 1], tsne2 = tsne.norm$Y[,2])
return(info.norm)
}

plot_tsne = function(info.norm){
tsne_eur3_pop = ggplot(info.norm, aes(x = tsne1, y = tsne2, colour = eurpop)) + 
  geom_point(alpha = 0.3) + theme_bw()

tsne_eur3_coh = ggplot(info.norm, aes(x = tsne1, y = tsne2, colour = cohort)) + 
  geom_point(alpha = 0.3) + theme_bw()

multiplot(tsne_eur3_pop, tsne_eur3_coh, cols = 2)
}
```

##data preprocessing: Adding factor for visualisation

```{r echo=FALSE, message=FALSE}
comb_pheno <- read_excel("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/git_proj_rvas/Comb_set_round2/ISKS_MGRB_2020/ISKS_AR_AD/Aug31_freeze/review_add/R2Q6_popstruc/Ethnicity.xlsx",
                         sheet = 1, col_types = c("list"))
comb_pheno <- as.data.frame(comb_pheno)
comb_pheno1 <- sapply(comb_pheno, unlist)
colnames(comb_pheno1) <- colnames(comb_pheno)
comb_pheno <- comb_pheno1
comb_pheno <- as.data.frame(comb_pheno, stringsAsFactors = F)
comb_pheno <- unique(comb_pheno)
comb_pheno <- comb_pheno[!is.na(comb_pheno$pid),]
comb_pheno$`age at dateExtracted` <- as.numeric(comb_pheno$`age at dateExtracted`)
comb_pheno$AgeatSarcoma <- as.numeric(comb_pheno$AgeatSarcoma)
comb_pheno$SubjectAgeCancer <- as.numeric(comb_pheno$SubjectAgeCancer)

#scores <- p_Data_noCH
scores = read.table("~/RVAS/comb_set_2020/pop_PCA/MGRB_ISKS_1000G_combset_pca.scores_clustered_rect_sam.tsv", header = T, sep = "\t", stringsAsFactors = F)

scores = scores[scores$superPopulation %in% c("ISKS", "RISC", "LIONS", "MGRB"),]
scores$Country <- comb_pheno[match(scores$sample, comb_pheno$pmn),64]
scores$Country <- ifelse(is.na(scores$Country), "Aus_MGRB", scores$Country)
scores$Country <- ifelse((!grepl("^[ABZ]",scores$sample) & (scores$Country %in% "Aus_MGRB")), "Unknown", scores$Country)

#scores = scores[scores$Country %nin% "Unknown",]

scores$superPopulation <- ifelse(scores$superPopulation %in% "MGRB", "MGRB", "ISKS")
scores$superPopulation <- ifelse(scores$superPopulation %in% "MGRB", "controls", "sarcoma")

scores_subPC1 = scores[scores$PC1 >= -0.2,]
 
##European data
scores_subPC1_Eur = scores_subPC1[scores_subPC1$pred.superPop %in% "EUR",]

```

####Analysis
##Exploratory Analysis of existing clustering after PC1 based filtering

#case control ratio
```{r echo=FALSE, message=FALSE}

cc_pop = case_cont_ratio_pop(scores_subPC1)
cc_eur = case_cont_ratio_eur(scores_subPC1_Eur)

```


#visualization (2D and 3D)
```{r echo=FALSE, message=FALSE, fig.align="center"}

##2D plots
plot_case_cont_2D(scores_subPC1)
plot_eur_case_cont_2D(scores_subPC1_Eur)

##3D plots
pop_3D = make_3Dplot_pop(scores_subPC1, theta = -120, phi = 0)
pop_3D = pop_3D + ggtitle("supPop")
eurpop_3D = make_3Dplot_eurpop(scores_subPC1_Eur, theta = -120, phi = 0)
eurpop_3D = eurpop_3D + ggtitle("EurPop")


pop_3D_cc = make_3Dplot_coh(scores_subPC1, theta = -120, phi = 0)
pop_3D_cc = pop_3D_cc + ggtitle("supPop")
eurpop_3D_cc = make_3Dplot_coh(scores_subPC1_Eur, theta = -120, phi = 0)
eurpop_3D_cc = eurpop_3D_cc + ggtitle("EurPop")

multiplot(pop_3D, pop_3D_cc, eurpop_3D, eurpop_3D_cc, cols = 2)
##tsne plot for Eurpop for better visualization
Eur_vis_inp = run_tsne_viz(scores_subPC1_Eur)
plot_tsne(Eur_vis_inp)

```


##Agglomerative clustering

```{r echo=FALSE, message=FALSE}

library(kableExtra)
##Uses first 4 PCs
##All populations : 5 superpopulations
scores_subPC1_clust = perf_clust(scores_subPC1, 5, "supPop")
scores_subPC1_clust$clust_code = paste("C", scores_subPC1_clust$cluster, sep = "_")
##case control ratio
clust_pop_cc = case_cont_ratio_clust(scores_subPC1_clust, "All")
#rownames(clust_pop_cc) = clust_pop_cc$cluster


##Eurpop : 5 european populations
scores_subPC1_clust_Eur = perf_clust(scores_subPC1_Eur, 5, "EurPop")
scores_subPC1_clust_Eur$clust_code = paste("C", scores_subPC1_clust_Eur$cluster, sep = "_")
##case control ratio
clust_eurpop_cc = case_cont_ratio_clust(scores_subPC1_clust_Eur, "EUR")

##print tables
clust_pop_cc %>%
  kbl(caption = "Super Populations") %>%
  kable_styling()

clust_eurpop_cc %>%
  kbl(caption = "European Populations") %>%
  kable_styling()

```


##Visualise unsupervised clusters

```{r echo=FALSE, message=FALSE}
pop_3D_clust = make_3Dplot_clust(scores_subPC1_clust, theta = -120, phi = 0)
pop_3D_clust = pop_3D_clust + ggtitle("supPop")
eurpop_3D_clust = make_3Dplot_clust(scores_subPC1_clust_Eur, theta = -120, phi = 0)
eurpop_3D_clust = eurpop_3D_clust + ggtitle("EurPop")

multiplot(pop_3D_clust, pop_3D_cc, eurpop_3D_clust, eurpop_3D_cc, cols = 2)

# viz_inp_clust = run_tsne_clust(scores_subPC1_clust_Eur)
# plot_tsne(viz_inp_clust)
# plot_tsne(Eur_vis_inp)

```


##Samples to remove

```{r}

rem_amr = scores_subPC1_clust[scores_subPC1_clust$clust_code %in% "C_2",]$rect_sam

rem_sas = scores_subPC1_clust[scores_subPC1_clust$clust_code %in% "C_3",]$rect_sam

rem_ibs = scores_subPC1_clust[scores_subPC1_clust_Eur$clust_code %in% "C_2",]$rect_sam


##Variant file used for SKAT
fil_tab <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/all_isksmgrb_skatinp_combset2020_clin_C3C4C5_NFE0002_AD_rm_dup_freeze.tsv", 
                      sep = "\t", header = T, stringsAsFactors = F)

fil_tab <- fil_tab[fil_tab$SAMPLE %in% scores_subPC1_clust$rect_sam,]


##chk_gene function

chk_gene = function(samp_inp){
  genes_outlier <- unique(fil_tab[fil_tab$SAMPLE %in% samp_inp,]$gene_symbol)
  Shelterin <- c("POT1", "TINF2", "TERF1", "TERF2", "TERF2IP", "SMARCAL1", "STAG3", "TIMELESS")
CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")
MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHD")
#table(genes_outlier %in% Shelterin)
  ##complex genes
print(genes_outlier[genes_outlier %in% Shelterin])  #"TIMELESS", "TERF2"; both C3
print(genes_outlier[genes_outlier %in% CEP_HAUS_core]) ##None
print(genes_outlier[genes_outlier %in% MPNST_pos])

##sample, genes, variants with mutations in complex
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% Shelterin,]$auto_call))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% Shelterin,]$SAMPLE))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% Shelterin,]$gene_symbol))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% CEP_HAUS_core,]$auto_call))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% CEP_HAUS_core,]$SAMPLE))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% CEP_HAUS_core,]$gene_symbol))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% MPNST_pos,]$auto_call))
print(table(fil_tab[fil_tab$SAMPLE %in% samp_inp & fil_tab$gene_symbol %in% MPNST_pos,]$gene_symbol))
}


chk_gene(rem_amr)
chk_gene(rem_sas)
chk_gene(rem_ibs)

##AMR C_2
##keep sample 2711, 1575, 1695
##can't keep 2429, its at the extremme
#scores_subPC1_clust$vis_ret = ifelse(scores_subPC1_clust$rect_sam %in% c("2429", "2711", "1575", "1695"), "Y", "N")
scores_subPC1_clust$vis_ret = ifelse(scores_subPC1_clust$rect_sam %in% c("2429"), "Y", "N")

ggplot(scores_subPC1_clust, aes(x=PC1, y=PC2, z=PC3, colour=vis_ret)) +
    axes_3D(theta = -120, phi = 0) +
    stat_3D(theta = -120, phi = 0, geom="point") +
    labs_3D(theta = -120, phi = 0, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 

##finalise removal
keep_amr_case = c("2711", "1575", "1695")

rem_amr_fin_ind = ifelse(rem_amr %in% keep_case | grepl("^[ABZ]", rem_amr), 0, 1)
rem_amr_fin = rem_amr[rem_amr_fin_ind == 1]

##SAS C_2

keep_sas_case = c("CR601", "LKCGP-P004101-269438-02-04-07-G1", "1285")

scores_subPC1_clust$vis_ret = ifelse(scores_subPC1_clust$rect_sam %in% keep_sas_case, "Y", "N")

ggplot(scores_subPC1_clust, aes(x=PC1, y=PC2, z=PC3, colour=vis_ret)) +
    axes_3D(theta = -120, phi = 0) +
    stat_3D(theta = -120, phi = 0, geom="point") +
    labs_3D(theta = -120, phi = 0, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 

##IBS C_2
keep_ibs_case = c("169", "LKCGP-T005701-242802-01-04-07-G1", "1368", "2522", "3540")
rem_ibs_case = c("LKCGP-T005701-242802-01-04-07-G1")

scores_subPC1_clust_Eur$vis_ret = ifelse(scores_subPC1_clust_Eur$rect_sam %in% keep_ibs_case, "Y", "N")
scores_subPC1_clust_Eur$vis_ret = ifelse(scores_subPC1_clust_Eur$rect_sam %in% "LKCGP-T005701-242802-01-04-07-G1", "Y", "N")

ggplot(scores_subPC1_clust_Eur, aes(x=PC1, y=PC2, z=PC3, colour=vis_ret)) +
    axes_3D(theta = -120, phi = 0) +
    stat_3D(theta = -120, phi = 0, geom="point") +
    labs_3D(theta = -120, phi = 0, 
            labs=c("PC1", "PC2", "PC3"), 
            angle=c(0,0,0),
            hjust=c(1,1,1), 
            vjust=c(2,2,-1)) +
    theme_void() 

#scores_subPC1_clust$vis_ret = ifelse(scores_subPC1_clust$rect_sam %nin% rem_amr_fin, "Y", "N")
scores_subPC1_clust_amr = scores_subPC1_clust[scores_subPC1_clust$rect_sam %nin% rem_amr_fin,]

amr_rm_clust = make_3Dplot_clust(scores_subPC1_clust_amr, theta = -120, phi = 0)
amr_rm_coh =make_3Dplot_coh(scores_subPC1_clust_amr, theta = -120, phi = 0)


multiplot(pop_3D_clust, pop_3D_cc, amr_rm_clust,  amr_rm_coh, cols = 2)


```

##Plots after removal of samples


