##Note that this analysis include singletons
##changes included on Aug-06-2019:
#added cohort_MAF computation feature; 
#input is better filtered compared to Aug4 input
##Aug12: SKAT made feasible for combined ISKS and RISC samples
#the rare variants are filtered using intra cohort_MAF filter that removes variants that are found in > 3 samples
#in either case of control(added after sep05 output)

.libPaths(c( "/home/shu/R/x86_64-redhat-linux-gnu-library/3.4", .libPaths() ) )
`%nin%` = Negate(`%in%`)

##Using D data structure

library(data.table)
library(VariantAnnotation)
library(stringr)
library(SKAT)
#library(githubinstall)
#githubinstall("SKATR")
#library(SKATR)
#library(CompQuadForm)

###get input files and process
##Experiments

fil_tab <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/all_isksmgrb_skatinp_combset2020_clin_C3C4C5_NFE0002_AD_rm_dup_freeze_rev_PC1_sub.tsv",
                      sep = "\t", stringsAsFactors = F, header = T) ##generated by R2Q6 script pca_pop*ISKS.R
fil_tab <- fil_tab[!is.na(fil_tab$SAMPLE),]
dim(fil_tab)

##Additive model
Ex_samp_id <- unique(fil_tab$SAMPLE)


#Ex_samp_id <- Ex_samp_id[Ex_samp_id %nin% dup_samp]

#Ex_var_id <- unique(fil_tab$VARIANT)

######get phenotype data to control for age and sex and PC's; 

QC2_dat <- read.table("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/Rep_set_variants/joint_calls_2019jan_2019nov.final_qc_output.tsv", 
                      header = T, sep = "\t", stringsAsFactors = F)
QC2_dat_pass <- QC2_dat[QC2_dat$passes_qc2 %in% "TRUE",]
QC2_dat_pass$isFemale <- ifelse(QC2_dat_pass$f_stat < 0.2, 1, 
                                ifelse(QC2_dat_pass$f_stat > 0.8, 0, 2))

##note p_Data_PC_comb does not change
#p_Data <- read.table("~/RVAS/ISKS_MGRB_gender_age_3665.tsv", header = T, sep = "\t")
#p_Data <- read.table("~/RVAS/comb_set_2020/pop_PCA/MGRB_ISKS_1000G_pca_combset.scores.tsv", header = T, sep = "\t", stringsAsFactors = F)
##for future QC with prob.NFE correction use the following file and adjust the column for gender from 39 to 53
##consolidating case control ratios using script pca_tsne_diagnostics.Rmd
#p_Data_noCH <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/rect_case_cont_ratio_PCA_clustered_jan2022.rds")
#p_Data_noCH <- readRDS("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_PCnew.rds") #change column names in Null model back to 54:55
p_Data_noCH <- read.delim("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_rev2_combset_pca.scores_clustered.tsv", sep = "\t", header = T, stringsAsFactors = F)
p_Data_noCH$rect_sam = p_Data_noCH$sample # the samples name are rectified at HAIL stage
p_Data_noCH <- p_Data_noCH[as.character(p_Data_noCH$rect_sam) %in% Ex_samp_id,]

Ex_samp_id <- Ex_samp_id[match(p_Data_noCH$rect_sam, Ex_samp_id)]
##filter out QC fail cases
fil_tab <- fil_tab[fil_tab$SAMPLE %in% Ex_samp_id,]

##Add gender information

p_Data_noCH$gender <- QC2_dat_pass[match(p_Data_noCH$rect_sam, QC2_dat_pass$new_sampleid), 21]
#write.table(p_Data_noCH, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/PID/pheno_data_SKAT_final_freeze.tsv",
#            sep = "\t", row.names = F, quote = F)

##Add exome count
##ref: Exome sequencing in amyotrophic lateral sclerosis implicates a novel gene, DNAJC7, encoding a heat-shock protein
##total exome count (summation of synonymous variants, benign missense variants, damaging missense variants and PTVs) 
tot_exome_count <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/all_isks_mgrb_total_exome_count_clingene_rnd3_freeze.tsv",
                              sep = "\t", header = T, stringsAsFactors = F)
p_Data_noCH$totex_count <- tot_exome_count[match(p_Data_noCH$rect_sam, tot_exome_count$Var1), 2]

#binary phenotype vector 
p_vec <- ifelse(!is.na(as.numeric(as.character(p_Data_noCH$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH$rect_sam)), 1, 0)

##Add EMMAX GRM
##Note: This is ancestry adjusted GRM, ancestry PCs must be added to null model
mat_GRM <- readRDS("~/RVAS/comb_set_2020/PC_relate/ldpruned/GRM_KIN_ISKS_MGRB.rds")
samp_all <- as.data.frame(colnames(mat_GRM))
#table(grepl("^[ABZ]", samp_all$Samp))
colnames(samp_all)[1] <- "Samp"
samp_all$Coh <- ifelse(grepl("^[ABZ]", samp_all$Samp), "MGRB", "ISKS")
rect_sam_dat <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/PID/ISKS_RISC_LIONS_final_freeze.tsv",
                           sep = "\t", header = T, stringsAsFactors = F)
samp_all$rect_sam <- rect_sam_dat[match(samp_all$Samp,rect_sam_dat$JCInputID),9]
rect_MGRB <- Ex_samp_id[grepl("^[ABZ]", Ex_samp_id)]
rect_ISKS <- Ex_samp_id[!grepl("^[ABZ]", Ex_samp_id)]
samp_all$rect_sam_MGRB <- rect_MGRB[match(samp_all$Samp, rect_MGRB)]
samp_all$rect_sam_ind <-  ifelse(!is.na(samp_all$rect_sam) & samp_all$rect_sam %in% rect_ISKS, 1,
                                 ifelse(!is.na(samp_all$rect_sam_MGRB), 1, 0))
coalesce2 <- function(...) {
  Reduce(function(x, y) {
    i <- which(is.na(x))
    x[i] <- y[i]
    x},
    list(...))
}
samp_all$coal <- coalesce2(samp_all$rect_sam,samp_all$rect_sam_MGRB)

mat_GRM_rect <- mat_GRM[which(samp_all$rect_sam_ind == 1), which(samp_all$rect_sam_ind == 1)]
colnames(mat_GRM_rect) <- samp_all[which(samp_all$rect_sam_ind == 1),]$coal
rownames(mat_GRM_rect) <- samp_all[which(samp_all$rect_sam_ind == 1),]$coal
mat_GRM_rect <- mat_GRM_rect[,p_Data_noCH$rect_sam]
mat_GRM_rect <- mat_GRM_rect[p_Data_noCH$rect_sam,]
#saveRDS(mat_GRM_rect, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/mat_GRM_rect.rds", compress = T)

##SKAT null function with customised covariate
SKAT_fun_null <- function(x=NULL, p_vec, GRM){
  if(is.null(x)){
    obj_N <- SKAT::SKAT_NULL_emmaX(p_vec ~ 1, K = GRM)
  }
  # else if(is.integer(x) || is.numeric()){
  else if(is.null(dim(x))){
    obj_N <- SKAT::SKAT_NULL_emmaX(p_vec ~ x, K = GRM)
  }
  else if(dim(x)[2] > 1){
    nul_for <- as.formula(paste("p_vec", paste(colnames(x), collapse = " + "), sep = " ~ "))
    obj_N <- SKAT::SKAT_NULL_emmaX(nul_for, data = p_Data_noCH, K = GRM)
    #  saveRDS(obj_N, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/Emmax_SKAT_NUll_obj.rds", compress = T)
    #  obj_N <- SKAT_Null_Model(nul_for, data = x, out_type="D")
  }
  return(obj_N)
}
null_d <- SKAT_fun_null(p_Data_noCH[,c(54:55,19:22)], p_vec, mat_GRM_rect)
##accounts for gender, total exome count, ancestry and genetic relatedness
#null_d <- SKAT_fun_null(p_Data_noCH[,c(55:56,19:22)], p_vec, mat_GRM_rect)

##SKAT function for SKAT, SKATBinary, SKATO, SKAT_ERA
SKAT_run_new <- function(geno_mat, gene, null_obj=NULL, p_vec, cust_weight = NULL, rho = NULL){
  
  pval_SKATburden <- SKAT::SKAT(t(geno_mat), null_obj, method = "Burden", weights = cust_weight)$p.value
  skat_pv <- cbind.data.frame("eg_ID" = gene, pval_SKATburden)
  return(skat_pv)
}

##Run SKAT on variants collapsed by gene

DT_skat_snv_str_pc123 <- list()
#genes <- unique(fil_tab$gene_symbol)

skat_pvals <- list()
skat_pvals_pc12 <- list()
skat_pvals_sex <- list()
skat_pvals_sex_pc12 <- list()
#    skat_pvals_age <- list()
#   skat_pvals_age_pc12 <- list()
#   skat_pvals_age_sex <- list()
#    skat_pvals_age_sex_pc12 <- list()

#for(i in 1:length(genes)){
# print(i)
#  print(genes[i])
##process genes in a loop

##parallel process genes
para_SKAT <- function(genes){
  ftemp_tab <- fil_tab[fil_tab$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  
  print(max(ftemp_tab$comb_score))
  if(dim(ftemp_tab) == 0 ){
    next
  }
  else{
    ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
    samp_vec <- list()
    for(m in 1:length(ftemp_tab_var_id)){
      sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
      sam_gene_gt <- unique(sam_gene_gt)
      if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
        # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
        vep_con <- unique(sam_gene_gt$vep_consequence)
        samp_vec_con <- list()
        for(k in 1:length(vep_con)){
          sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
          sam_gene_gt_con <- unique(sam_gene_gt_con)
          maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          ##genotype matrix  
          sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
          sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt_con$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
          samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                 unique(sam_gene_gt_con$gene_symbol), 
                                 unique(sam_gene_gt_con$vep_consequence), 
                                 unique(as.character(sam_gene_gt_con$auto_call)),
                                 as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                 as.numeric(maf_vec_cont), sam10)
        }
        #  vep_cod <- sapply(str_extract_all(sam_gene_gt$vep_Codons, "[A-Z]+"), paste, collapse= '_')
        #  ref_alt <- unlist(lapply(strsplit(sam_gene_gt$VARIANT, split = ":"), function(x) paste(x[3], x[4], sep = "_")))
        #  sam_gene_gt <- sam_gene_gt[which(vep_cod %in% ref_alt),]
        
      }
      else{
        ##compute cohort specific MAF
        maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                     grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
        ##genotype matrix  
        sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
        sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt$SAMPLE, 1, 0)
        sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
        ##account for discrepancy in vep_consequence
        #vep_con <- names(sort(table(unlist(strsplit(sam_gene_gt$vep_consequence, split = "&"))),decreasing=TRUE)[1])
        # sam_gene_gt$vep_consequence <- vep_con
        
        
        samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                           unique(sam_gene_gt$gene_symbol), 
                           unique(sam_gene_gt$vep_consequence), 
                           unique(as.character(sam_gene_gt$auto_call)),
                           as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                           as.numeric(maf_vec_cont), sam10)
      }
    }
    samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
    colnames(samp_vec_mat_uni) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                    "coh_MAF_cont", Ex_samp_id)
    if(exists("samp_vec_con")){
      samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
      colnames(samp_vec_mat_con) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id)
      samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
    }else{
      samp_vec_mat <- samp_vec_mat_uni
    }
    print(dim(samp_vec_mat))
    ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
    ##changed to uniform intra_cohort_MAF filter
    ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
    #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
    #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    ##After removal of AMR and AFR entire cohort MAF filter to 3/(4788*2)##latest Jan.10.2022
    samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) >= 0.00035 | 
                                     as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) >= 0.00035),]
    #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    
    if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
      #skat_pvals <- NULL
      #skat_pvals_pc12 <- NULL
      #skat_pvals_sex <- NULL
      skat_pvals_sex_pc12 <- NULL
      # skat_pvals_age[[i]] <- NULL
      # skat_pvals_age_pc12[[i]] <- NULL
      # skat_pvals_age_sex[[i]] <- NULL
      # skat_pvals_age_sex_pc12[[i]] <- NULL
      #   break
      next
    }
    ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
    
    else {
      gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
      gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
      class(gene_mat) <- "numeric"
      ##for strictly burden test set rho = 1, else set rho = 0
      skat_pvals_sex_pc12 <- SKAT_run_new(geno_mat = gene_mat, null_obj =  null_d, gene = genes, p_vec, cust_weight = gene_mat_comb, rho = 1)
      
    }
    
  } ##end of nested else loop
  
  return(skat_pvals_sex_pc12)
}##end of gene loop ; i loop   

##Parallelised SKAT  
library(doParallel)
library(doMC)
registerDoMC(30)

res20 <- list()
genes <- unique(fil_tab$gene_symbol)
system.time(res20 <- foreach(i=1:length(genes), .errorhandling = 'remove') %dopar% 
{para_SKAT(genes[i])})

saveRDS(res20, file = "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/SKAT/Exome_skat_para_result_isks_totex_count_PC1234_subPC1_EMMAX_Jan10_ccrect_PCnew.rds", compress = T)
