###APcluster + popstruc for rare variants in multiethnic setting
##only test for genes identified at discovery stage as a validation

.libPaths(c( "/home/shu/R/x86_64-redhat-linux-gnu-library/3.4", .libPaths() ) )
`%nin%` = Negate(`%in%`)

library(data.table)
library(VariantAnnotation)
library(stringr)
library(SKAT)

##Data processing

fil_tab <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/all_isksmgrb_skatinp_combset2020_clin_C3C4C5_NFE0002_AD_rm_dup_freeze_rev_PC1_sub.tsv",
                      sep = "\t", stringsAsFactors = F, header = T) ##generated by R2Q6 script pca_pop*ISKS.R
fil_tab <- fil_tab[!is.na(fil_tab$SAMPLE),]
dim(fil_tab)




##Additive model
Ex_samp_id <- unique(fil_tab$SAMPLE)


#Ex_samp_id <- Ex_samp_id[Ex_samp_id %nin% dup_samp]

#Ex_var_id <- unique(fil_tab$VARIANT)

######get phenotype data to control for age and sex and PC's; 

QC2_dat <- read.table("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/Rep_set_variants/joint_calls_2019jan_2019nov.final_qc_output.tsv", 
                      header = T, sep = "\t", stringsAsFactors = F)
QC2_dat_pass <- QC2_dat[QC2_dat$passes_qc2 %in% "TRUE",]
QC2_dat_pass$isFemale <- ifelse(QC2_dat_pass$f_stat < 0.2, 1, 
                                ifelse(QC2_dat_pass$f_stat > 0.8, 0, 2))
##note p_Data_PC_comb does not change
#p_Data <- read.table("~/RVAS/ISKS_MGRB_gender_age_3665.tsv", header = T, sep = "\t")
#p_Data <- read.table("~/RVAS/comb_set_2020/pop_PCA/MGRB_ISKS_1000G_pca_combset.scores.tsv", header = T, sep = "\t", stringsAsFactors = F)
##for future QC with prob.NFE correction use the following file and adjust the column for gender from 39 to 53

#p_Data_noCH <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/rect_case_cont_ratio_PCA_clustered_jan2022.rds")
#p_Data_noCH <- readRDS("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_PCnew.rds") #change column names in Null model back to 54:55
p_Data_noCH = read.delim("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_rev2_combset_pca.scores_clustered.tsv", header = T, stringsAsFactors = F)
p_Data_noCH$rect_sam = p_Data_noCH$sample # the samples name are rectified at HAIL stage
p_Data_noCH <- p_Data_noCH[as.character(p_Data_noCH$rect_sam) %in% Ex_samp_id,]
##drop MGRB sample BAAUD; not in latest call
#samp_ID_match <- samp_ID_match[grep("BAAUD", samp_ID_match, invert = T)]
#p_Data_noCH <- p_Data_noCH[match(samp_ID_match, p_Data_noCH$sample),]
Ex_samp_id <- Ex_samp_id[match(p_Data_noCH$rect_sam, Ex_samp_id)]
##filter out QC fail cases
fil_tab <- fil_tab[fil_tab$SAMPLE %in% Ex_samp_id,]

##Add gender information

p_Data_noCH$gender <- QC2_dat_pass[match(p_Data_noCH$rect_sam, QC2_dat_pass$new_sampleid), 21]
#write.table(p_Data_noCH, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/PID/pheno_data_SKAT_final_freeze.tsv",
#            sep = "\t", row.names = F, quote = F)

##Add exome count
##ref: Exome sequencing in amyotrophic lateral sclerosis implicates a novel gene, DNAJC7, encoding a heat-shock protein
##total exome count (summation of synonymous variants, benign missense variants, damaging missense variants and PTVs) 
tot_exome_count <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/all_isks_mgrb_total_exome_count_clingene_rnd3_freeze.tsv",
                              sep = "\t", header = T, stringsAsFactors = F)
p_Data_noCH$totex_count <- tot_exome_count[match(p_Data_noCH$rect_sam, tot_exome_count$Var1), 2]

p_Data_noCH$totex_count = as.numeric(p_Data_noCH$totex_count)

fil_tab$pop = p_Data_noCH[match(fil_tab$SAMPLE, p_Data_noCH$rect_sam), 39]

##run only for discovered complexes


#binary phenotype vector 
p_vec <- ifelse(!is.na(as.numeric(as.character(p_Data_noCH$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH$rect_sam)), 1, 0)


##SKAT null function with customised covariate
SKAT_fun_null <- function(x=NULL, p_vec_pop){
  if(is.null(x)){
    obj_N <- SKAT::SKAT_Null_Model(p_vec_pop ~ 1,out_type="D")
  }
  # else if(is.integer(x) || is.numeric()){
  else if(is.null(dim(x))){
    obj_N <- SKAT::SKAT_Null_Model(p_vec_pop ~ x,out_type="D")
  }
  else if(dim(x)[2] > 1){
    nul_for <- as.formula(paste("p_vec_pop", paste(colnames(x), collapse = " + "), sep = " ~ "))
   # obj_N <- SKAT::SKAT_Null_Model(nul_for, data = p_Data_noCH_sub, out_type="D")
      obj_N <- SKAT::SKAT_Null_Model(nul_for, data = x, out_type="D")
  }
  return(obj_N)
}

##SKAT function for SKAT, SKATBinary, SKATO, SKAT_ERA
SKAT_run <- function(geno_mat, gene, x=NULL, p_vec_pop, cust_weight = NULL, rho = NULL){
  
  if(dim(geno_mat)[2] > 1 ){
    null_d <- SKAT_fun_null(x, p_vec_pop)
    # pval_SKAT <- SKAT::SKAT(t(geno_mat), null_d, weights = cust_weight, r.corr = rho)$p.value
    pval_SKATbin <- SKAT::SKATBinary(t(geno_mat), null_d, method = "Burden", weights = cust_weight)$p.value
    skat_pv <- cbind.data.frame("eg_ID" = gene, pval_SKATbin)
    return(skat_pv)
  }
  else if(dim(geno_mat)[2] == 1){
    null_d <- SKAT::SKAT_fun_null(x, p_vec_pop)
    pval_SKATbin <- SKAT::SKATBinary(t(geno_mat), null_d, method = "Burden", weights = cust_weight)$p.value
    skat_pv <- cbind.data.frame("eg_ID" = gene, pval_SKATbin)
    return(skat_pv)
  }
}

skat_pvals_sex_pc12 = list()
##parallel process genes
para_SKAT_pop <- function(genes){
  pop = unique(fil_tab$pop)
  for(j in 1:length(pop)){
    fil_tab_sub = fil_tab[fil_tab$pop %in% pop[j],]
    print(pop[j])
    Ex_samp_id_pop = unique(fil_tab_sub$SAMPLE)
    p_Data_noCH_sub = p_Data_noCH[p_Data_noCH$rect_sam %in% Ex_samp_id_pop,]
    Ex_samp_id_pop <- Ex_samp_id_pop[match(p_Data_noCH_sub$rect_sam, Ex_samp_id_pop)]
    p_Data_noCH_sub = p_Data_noCH_sub[match(Ex_samp_id_pop, p_Data_noCH_sub$rect_sam),]
    p_vec_pop = ifelse(!is.na(as.numeric(as.character(p_Data_noCH_sub$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH_sub$rect_sam)), 1, 0)

      
  ftemp_tab <- fil_tab_sub[fil_tab_sub$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  
  print(max(ftemp_tab$comb_score))
  if(dim(ftemp_tab) == 0 ){
    skat_pvals_sex_pc12[[j]] <- cbind.data.frame("eg_ID" = genes, "pval_SKATbin" = 1) 
    next
  }
  else{
    ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
    samp_vec <- list()
    for(m in 1:length(ftemp_tab_var_id)){
      sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
      sam_gene_gt <- unique(sam_gene_gt)
      if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
        # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
        vep_con <- unique(sam_gene_gt$vep_consequence)
        samp_vec_con <- list()
        for(k in 1:length(vep_con)){
          sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
          sam_gene_gt_con <- unique(sam_gene_gt_con)
          maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec_pop))
          maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec_pop))
          ##genotype matrix  
          sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
          sam10 <- ifelse(Ex_samp_id_pop %in% sam_gene_gt_con$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
          samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                 unique(sam_gene_gt_con$gene_symbol), 
                                 unique(sam_gene_gt_con$vep_consequence), 
                                 unique(as.character(sam_gene_gt_con$auto_call)),
                                 as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                 as.numeric(maf_vec_cont), sam10)
        }
        
      }
      else{
        ##compute cohort specific MAF
        maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec_pop))
        maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                     grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec_pop))
        #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
        ##genotype matrix  
        sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
        sam10 <- ifelse(Ex_samp_id_pop %in% sam_gene_gt$SAMPLE, 1, 0)
        sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
        
        
        samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                           unique(sam_gene_gt$gene_symbol), 
                           unique(sam_gene_gt$vep_consequence), 
                           unique(as.character(sam_gene_gt$auto_call)),
                           as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                           as.numeric(maf_vec_cont), sam10)
      }
    }
    samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
    colnames(samp_vec_mat_uni) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                    "coh_MAF_cont", Ex_samp_id_pop)
    if(exists("samp_vec_con")){
      samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
      colnames(samp_vec_mat_con) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id_pop)
      samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
    }else{
      samp_vec_mat <- samp_vec_mat_uni
    }
    print(dim(samp_vec_mat))
    ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
    ##changed to uniform intra_cohort_MAF filter
    ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
    #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
    #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    ##After removal of AMR and AFR entire cohort MAF filter to 3/(4788*2)##latest Jan.10.2022
    samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) > 3/(2*length(p_vec_pop)) | 
                                     as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) > 3/(2*length(p_vec_pop))),]
    #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    
    if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
      #skat_pvals <- NULL
      #skat_pvals_pc12 <- NULL
      #skat_pvals_sex <- NULL
      skat_pvals_sex_pc12[[j]] <- cbind.data.frame("eg_ID" = genes, "pval_SKATbin" = 1)
      # skat_pvals_age[[i]] <- NULL
      # skat_pvals_age_pc12[[i]] <- NULL
      # skat_pvals_age_sex[[i]] <- NULL
      # skat_pvals_age_sex_pc12[[i]] <- NULL
      #   break
      next
    }
    ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
    
    else {
      gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
      gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
      class(gene_mat) <- "numeric"
      print(dim(gene_mat))
      print(dim(p_Data_noCH_sub))
      print(table(p_vec_pop))
      ##for strictly burden test set rho = 1, else set rho = 0
      skat_pvals_sex_pc12[[j]] <- SKAT_run(geno_mat = gene_mat, x=p_Data_noCH_sub[,c(54:55)], gene = genes, p_vec_pop, cust_weight = gene_mat_comb, rho = 1)
      
    }
    
  } ##end of nested else loop
 # return(skat_pvals_sex_pc12)
}##end of gene loop ; i loop   
  names(skat_pvals_sex_pc12) = pop
  skat_pvals_sex_pc12_df = melt(skat_pvals_sex_pc12)
  skat_pvals_sex_pc12_df = skat_pvals_sex_pc12_df[,c(1,3,4)]
  colnames(skat_pvals_sex_pc12_df) = c("symbol", "p_value", "pop")
  return(skat_pvals_sex_pc12_df)
}

##Parallelised SKAT for complex genes
library(doParallel)
library(doMC)
registerDoMC(30)

Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")

CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")

MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")

all_genes = unique(c(Shelterin, CEP_HAUS_core, MPNST_pos, "TP53", "BRCA2", "BRCA1", "PALB2"))

fil_tab_sub = fil_tab[fil_tab$gene_symbol %in% all_genes,]


res20 <- list()
#genes <- unique(fil_tab$gene_symbol)
genes = all_genes
system.time(res20 <- foreach(i=1:length(genes), .errorhandling = 'remove') %dopar% 
{para_SKAT_pop(genes[i])})


res20_df = melt(res20)
##########################################
##for complexes
##parallel process complexes
para_SKAT_cpx <- function(genes, cpx_name){
  
  #pop = unique(fil_tab$pop)
 # for(j in 1:length(pop)){
    pop = "EUR"
    fil_tab_sub = fil_tab[fil_tab$pop %in% pop,]
    print(pop)
    Ex_samp_id_pop = unique(fil_tab_sub$SAMPLE)
    p_Data_noCH_sub = p_Data_noCH[p_Data_noCH$rect_sam %in% Ex_samp_id_pop,]
    Ex_samp_id_pop <- Ex_samp_id_pop[match(p_Data_noCH_sub$rect_sam, Ex_samp_id_pop)]
    p_Data_noCH_sub = p_Data_noCH_sub[match(Ex_samp_id_pop, p_Data_noCH_sub$rect_sam),]
    p_vec_pop = ifelse(!is.na(as.numeric(as.character(p_Data_noCH_sub$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH_sub$rect_sam)), 1, 0)
    
  
  ftemp_tab <- fil_tab_sub[fil_tab_sub$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  
  print(max(ftemp_tab$comb_score))
  if(dim(ftemp_tab) == 0 ){
    skat_pvals_sex_pc12[[j]] <- cbind.data.frame("eg_ID" = genes, "pval_SKATbin" = 1)
    next
  }
  else{
    ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
    samp_vec <- list()
    for(m in 1:length(ftemp_tab_var_id)){
      sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
      sam_gene_gt <- unique(sam_gene_gt)
      if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
        # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
        vep_con <- unique(sam_gene_gt$vep_consequence)
        samp_vec_con <- list()
        for(k in 1:length(vep_con)){
          sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
          sam_gene_gt_con <- unique(sam_gene_gt_con)
          maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec_pop))
          maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec_pop))
          ##genotype matrix  
          sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
          sam10 <- ifelse(Ex_samp_id_pop %in% sam_gene_gt_con$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
          samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                 unique(sam_gene_gt_con$gene_symbol), 
                                 unique(sam_gene_gt_con$vep_consequence), 
                                 unique(as.character(sam_gene_gt_con$auto_call)),
                                 as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                 as.numeric(maf_vec_cont), sam10)
        }
        #  vep_cod <- sapply(str_extract_all(sam_gene_gt$vep_Codons, "[A-Z]+"), paste, collapse= '_')
        #  ref_alt <- unlist(lapply(strsplit(sam_gene_gt$VARIANT, split = ":"), function(x) paste(x[3], x[4], sep = "_")))
        #  sam_gene_gt <- sam_gene_gt[which(vep_cod %in% ref_alt),]
        
      }
      else{
        ##compute cohort specific MAF
        maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec_pop))
        maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                     grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec_pop))
        #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
        ##genotype matrix  
        sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
        sam10 <- ifelse(Ex_samp_id_pop %in% sam_gene_gt$SAMPLE, 1, 0)
        sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
        ##account for discrepancy in vep_consequence
        #vep_con <- names(sort(table(unlist(strsplit(sam_gene_gt$vep_consequence, split = "&"))),decreasing=TRUE)[1])
        # sam_gene_gt$vep_consequence <- vep_con
        
        
        samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                           unique(sam_gene_gt$gene_symbol), 
                           unique(sam_gene_gt$vep_consequence), 
                           unique(as.character(sam_gene_gt$auto_call)),
                           as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                           as.numeric(maf_vec_cont), sam10)
      }
    }
    samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
    colnames(samp_vec_mat_uni) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                    "coh_MAF_cont", Ex_samp_id_pop)
    if(exists("samp_vec_con")){
      samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
      colnames(samp_vec_mat_con) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id_pop)
      samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
    }else{
      samp_vec_mat <- samp_vec_mat_uni
    }
    print(dim(samp_vec_mat))
    ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
    ##changed to uniform intra_cohort_MAF filter
    ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
    #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
    #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) >= 0.00035 | 
                                     as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) >= 0.00035),]
    #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    
    if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
    
      skat_pvals_sex_pc12[[j]] <- cbind.data.frame("eg_ID" = genes, "pval_SKATbin" = 1)
      # skat_pvals_age[[i]] <- NULL
      # skat_pvals_age_pc12[[i]] <- NULL
      # skat_pvals_age_sex[[i]] <- NULL
      # skat_pvals_age_sex_pc12[[i]] <- NULL
      #   break
      next
    }
    ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
    
    else {
      gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
      gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
      class(gene_mat) <- "numeric"
        ##gender + totex 
      skat_pvals_sex_pc12 <- SKAT_run(geno_mat = gene_mat, gene = cpx_name, x=p_Data_noCH_sub[,c(54:55)], p_vec_pop, cust_weight = gene_mat_comb, rho = 1)
      ##gender + totex + pc1:pc4
     # skat_pvals_sex_pc12 <- SKAT_run(geno_mat = gene_mat, gene = cpx_name, x=p_Data_noCH_sub[,c(54:55,19:22)], p_vec_pop, cust_weight = gene_mat_comb, rho = 1)
    }
    
  } ##end of nested else loop
  return(skat_pvals_sex_pc12)
  
}##end of gene loop ; i loop   
  


Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")
Sarcoma_genes <- c("TP53", "NF1", "SDHA", "SDHB", "SDHD")
CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")

Breast_cancer_genes = c("BRCA2", "BRCA1", "PALB2")

MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")

cpx_list <- list(Shelterin, CEP_HAUS_core, Breast_cancer_genes, Sarcoma_genes, MPNST_pos)
names(cpx_list) <- c("Shelterin", "Centrosome", "BRCA_genes", "SARC_genes", "MPNST")

##Parallelised SKAT  
library(doParallel)
library(doMC)
registerDoMC(30)


res20_cpx <- list()
#genes <- unique(fil_tab$gene_symbol)
#genes <- names(cpx_list)
system.time(res20_cpx <- foreach(i=1:length(cpx_list), .errorhandling = 'remove') %dopar% 
{para_SKAT_cpx(cpx_list[[i]], names(cpx_list)[i])})

res20_cpx_df = do.call("rbind.data.frame", res20_cpx)
res20_cpx_df$p.adj = p.adjust(res20_cpx_df$pval_SKATbin)


##try this with PCs for within Europe population
res20_cpx_pc <- list()

system.time(res20_cpx_pc <- foreach(i=1:length(cpx_list), .errorhandling = 'remove') %dopar% 
{para_SKAT_cpx(cpx_list[[i]], names(cpx_list)[i])})


res20_cpx_pc_df = do.call("rbind.data.frame", res20_cpx_pc)
res20_cpx_pc_df$p.adj = p.adjust(res20_cpx_pc_df$pval_SKATbin)

####remove variants with high frequency in admixture and non_EUR populations
##extract variants fro complexes; use dbNSFP to obtain AMR, AFR, SAS, EAS frequencies from gnomad

fil_tab_cpx = fil_tab[fil_tab$gene_symbol %in% c(Shelterin, CEP_HAUS_core, Breast_cancer_genes, Sarcoma_genes, MPNST_pos),2]

fil_tab_cpx_df = do.call("rbind.data.frame", lapply(strsplit(fil_tab_cpx, split = ":"), function(x)cbind.data.frame("chr" = x[1], "pos" = x[2],
                                                                                        "ref" = x[3], "alt" = x[4])))
fil_tab_cpx_df = unique(fil_tab_cpx_df)
##for dbnsfp run
write.table(fil_tab_cpx_df, "~/RVAS/dbNSFP_4.1a/isks_cpx_var.in", sep = "\t", row.names = F, quote = F)

##dbnsfp output; check if any of the variants have higher than 0.001 AF across any other population
##indels were excluded as dbnsfp is only for coding snvs

dbnsfp_cpx = read.delim("~/RVAS/dbNSFP_4.1a/isks_cpx_var_dbnsfp_out_fin.tsv", sep = "\t", header = T, stringsAsFactors = F)
for(i in 6:9){
dbnsfp_cpx[,i] <- as.numeric(dbnsfp_cpx[,i])
}
dbnsfp_cpx[is.na(dbnsfp_cpx)] = 0
dbnsfp_cpx =  unique(dbnsfp_cpx)
rownames(dbnsfp_cpx) = gsub(" ", "", apply(dbnsfp_cpx[ ,1:4] , 1 , paste, collapse = ":" ))

dbnsfp_cpx_ind_01 = apply(dbnsfp_cpx[,6:9], 1, function(x)ifelse(max(x) >= 0.001, 1, 0))
dbnsfp_cpx_samp_01 = names(dbnsfp_cpx_ind_01[dbnsfp_cpx_ind_01 == 1])
s1 = fil_tab[fil_tab$VARIANT %in% dbnsfp_cpx_samp_01,c(1:3,9,127:128,131)]
#fil_tab[fil_tab$SAMPLE %in% s1$SAMPLE & fil_tab$gene_symbol %in% c(Shelterin, CEP_HAUS_core, Breast_cancer_genes, Sarcoma_genes, MPNST_pos),c(1,9,127,131)]
##remove s1 variants from variant set


################
##use fil_tab and PCs after removal of samples using centroid method
fil_tab <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/SKAT/all_isksmgrb_skatinp_combset2020_clin_C3C4C5_NFE0002_AD_rm_dup_freeze_rev_PC1_sub.tsv",
                      sep = "\t", stringsAsFactors = F, header = T) ##generated by R2Q6 script pca_pop*ISKS.R
fil_tab <- fil_tab[!is.na(fil_tab$SAMPLE),]
dim(fil_tab)

##Additive model
Ex_samp_id <- unique(fil_tab$SAMPLE)

######get phenotype data to control for age and sex and PC's; 

QC2_dat <- read.table("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/Rep_set_variants/joint_calls_2019jan_2019nov.final_qc_output.tsv", 
                      header = T, sep = "\t", stringsAsFactors = F)
QC2_dat_pass <- QC2_dat[QC2_dat$passes_qc2 %in% "TRUE",]
QC2_dat_pass$isFemale <- ifelse(QC2_dat_pass$f_stat < 0.2, 1, 
                                ifelse(QC2_dat_pass$f_stat > 0.8, 0, 2))
##for future QC with prob.NFE correction use the following file and adjust the column for gender from 39 to 53

#p_Data_noCH <- readRDS("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_2/rect_case_cont_ratio_PCA_clustered_jan2022.rds")
p_Data_noCH <- readRDS("~/RVAS/comb_set_2020/pop_PCA/review_2/MGRB_ISKS_1000G_rect_PCnew.rds") #change column names in Null model back to 54:55
p_Data_noCH$rect_sam = p_Data_noCH$sample # the samples name are rectified at HAIL stage
p_Data_noCH <- p_Data_noCH[as.character(p_Data_noCH$rect_sam) %in% Ex_samp_id,]
##drop MGRB sample BAAUD; not in latest call
#samp_ID_match <- samp_ID_match[grep("BAAUD", samp_ID_match, invert = T)]
#p_Data_noCH <- p_Data_noCH[match(samp_ID_match, p_Data_noCH$sample),]
Ex_samp_id <- Ex_samp_id[match(p_Data_noCH$rect_sam, Ex_samp_id)]
##filter out QC fail cases
fil_tab <- fil_tab[fil_tab$SAMPLE %in% Ex_samp_id,]

p_Data_noCH$gender <- QC2_dat_pass[match(p_Data_noCH$rect_sam, QC2_dat_pass$new_sampleid), 21]
#write.table(p_Data_noCH, "~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/PID/pheno_data_SKAT_final_freeze.tsv",
#            sep = "\t", row.names = F, quote = F)

##Add exome count
##ref: Exome sequencing in amyotrophic lateral sclerosis implicates a novel gene, DNAJC7, encoding a heat-shock protein
##total exome count (summation of synonymous variants, benign missense variants, damaging missense variants and PTVs) 
tot_exome_count <- read.delim("~/RVAS/shard_sub_tier3/DT_sheet/EXOME_isks_risc/test/comb_set_2020/ISKS_AR_AD/review_doc/all_isks_mgrb_total_exome_count_clingene_rnd3_freeze.tsv",
                              sep = "\t", header = T, stringsAsFactors = F)
p_Data_noCH$totex_count <- tot_exome_count[match(p_Data_noCH$rect_sam, tot_exome_count$Var1), 2]

#binary phenotype vector 
p_vec <- ifelse(!is.na(as.numeric(as.character(p_Data_noCH$rect_sam))) | grepl("^CR|^LK",as.character(p_Data_noCH$rect_sam)), 1, 0)


##parallel process genes in discovered complex after removal of popmax variant greater than 0.001
para_SKAT_var <- function(genes, rm_var_file){
  ftemp_tab <- fil_tab[fil_tab$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  ftemp_tab <- ftemp_tab[ftemp_tab$VARIANT %nin% rm_var_file,]
    
    
  print(max(ftemp_tab$comb_score))
  if(dim(ftemp_tab) == 0 ){
    next
  }
  else{
    ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
    samp_vec <- list()
    for(m in 1:length(ftemp_tab_var_id)){
      sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
      sam_gene_gt <- unique(sam_gene_gt)
      if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
        # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
        vep_con <- unique(sam_gene_gt$vep_consequence)
        samp_vec_con <- list()
        for(k in 1:length(vep_con)){
          sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
          sam_gene_gt_con <- unique(sam_gene_gt_con)
          maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
          ##genotype matrix  
          sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
          sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt_con$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
          samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                 unique(sam_gene_gt_con$gene_symbol), 
                                 unique(sam_gene_gt_con$vep_consequence), 
                                 unique(as.character(sam_gene_gt_con$auto_call)),
                                 as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                 as.numeric(maf_vec_cont), sam10)
        }
        #  vep_cod <- sapply(str_extract_all(sam_gene_gt$vep_Codons, "[A-Z]+"), paste, collapse= '_')
        #  ref_alt <- unlist(lapply(strsplit(sam_gene_gt$VARIANT, split = ":"), function(x) paste(x[3], x[4], sep = "_")))
        #  sam_gene_gt <- sam_gene_gt[which(vep_cod %in% ref_alt),]
        
      }
      else{
        ##compute cohort specific MAF
        maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                     grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
        #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
        ##genotype matrix  
        sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
        sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt$SAMPLE, 1, 0)
        sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
        ##account for discrepancy in vep_consequence
        #vep_con <- names(sort(table(unlist(strsplit(sam_gene_gt$vep_consequence, split = "&"))),decreasing=TRUE)[1])
        # sam_gene_gt$vep_consequence <- vep_con
        
        
        samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                           unique(sam_gene_gt$gene_symbol), 
                           unique(sam_gene_gt$vep_consequence), 
                           unique(as.character(sam_gene_gt$auto_call)),
                           as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                           as.numeric(maf_vec_cont), sam10)
      }
    }
    samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
    colnames(samp_vec_mat_uni) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                    "coh_MAF_cont", Ex_samp_id)
    if(exists("samp_vec_con")){
      samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
      colnames(samp_vec_mat_con) <- c("VARIANT", "gene_symbol", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id)
      samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
    }else{
      samp_vec_mat <- samp_vec_mat_uni
    }
    print(dim(samp_vec_mat))
    ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
    ##changed to uniform intra_cohort_MAF filter
    ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
    #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
    #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) >= 0.00035 | 
                                     as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) >= 0.00035),]
    #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
    
    if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
      skat_pvals_sex_pc12 <- NULL
      next
    }
    ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
    
    else {
      gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
      gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
      class(gene_mat) <- "numeric"
      skat_pvals_sex_pc12 <- SKAT_run(geno_mat = gene_mat, gene = genes, x=p_Data_noCH[,c(55:56,19:22)], p_vec, cust_weight = gene_mat_comb, rho = 1)
      
    }
    
  } ##end of nested else loop
  # DT_skat_snv_str_pc123 <- list(skat_pvals, skat_pvals_pc12, skat_pvals_sex, skat_pvals_sex_pc12)
  # return(DT_skat_snv_str_pc123)
  return(skat_pvals_sex_pc12)
}##end of gene loop ; i loop   

##Parallelised SKAT  
library(doParallel)
library(doMC)
registerDoMC(30)

Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")

CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")

MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")

all_genes = unique(c(Shelterin, CEP_HAUS_core, MPNST_pos, "TP53", "BRCA2", "BRCA1", "PALB2"))

res20_var <- list()
#genes <- unique(fil_tab$gene_symbol)
system.time(res20_var <- foreach(i=1:length(all_genes), .errorhandling = 'remove') %dopar% 
{para_SKAT_var(all_genes[i], dbnsfp_cpx_samp_01)}) ##from POPMAX : line 454

res20_var_df = do.call("rbind.data.frame", res20_var)

##########################################
##for complexes after removal of POPMAX variant > 0.001, with new PCs
##parallel process complexes
para_SKAT_cpx_var <- function(genes, cpx_name, rm_var_file){
  skat_pvals_sex_pc12 = list()
  
  ftemp_tab <- fil_tab[fil_tab$gene_symbol %in% genes,]
  ftemp_tab <- ftemp_tab[ftemp_tab$comb_score >= 5.6 & as.numeric(ftemp_tab$VAF) >= 0.35, ]
  ftemp_tab <- ftemp_tab[ftemp_tab$VARIANT %nin% rm_var_file,]
 
    print(max(ftemp_tab$comb_score))
    if(dim(ftemp_tab) == 0 ){
      next
    }
    else{
      ftemp_tab_var_id <- unique(ftemp_tab$VARIANT)
      samp_vec <- list()
      for(m in 1:length(ftemp_tab_var_id)){
        sam_gene_gt <- ftemp_tab[ftemp_tab$VARIANT %in% ftemp_tab_var_id[m],][,c(1:3,9,11,127:128)]
        sam_gene_gt <- unique(sam_gene_gt)
        if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1){
          # if(dim(sam_gene_gt)[1] > 1 & length(unique(sam_gene_gt$vep_consequence)) > 1 & length(unique(sam_gene_gt$comb_score)) > 1){
          vep_con <- unique(sam_gene_gt$vep_consequence)
          samp_vec_con <- list()
          for(k in 1:length(vep_con)){
            sam_gene_gt_con <- sam_gene_gt[sam_gene_gt$vep_consequence %in% vep_con[k],]
            sam_gene_gt_con <- unique(sam_gene_gt_con)
            maf_vec_cont <- sum(dim(sam_gene_gt_con[grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
            maf_vec_case <- sum(dim(sam_gene_gt_con[!grepl("^[ABZ]",sam_gene_gt_con$SAMPLE),])[1])/(2*length(p_vec))
            ##genotype matrix  
            sam_gene_gt_con$add_mod <- as.numeric(sam_gene_gt_con$GT)
            sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt_con$SAMPLE, 1, 0)
            sam10[which(sam10 != 0)] <- sam_gene_gt_con$add_mod ##additive model
            samp_vec_con[[k]] <- c(unique(sam_gene_gt_con$VARIANT),
                                   unique(sam_gene_gt_con$gene_symbol), 
                                   unique(sam_gene_gt_con$vep_consequence), 
                                   unique(as.character(sam_gene_gt_con$auto_call)),
                                   as.numeric(unique(sam_gene_gt_con$comb_score)), as.numeric(maf_vec_case),
                                   as.numeric(maf_vec_cont), sam10)
          }
          #  vep_cod <- sapply(str_extract_all(sam_gene_gt$vep_Codons, "[A-Z]+"), paste, collapse= '_')
          #  ref_alt <- unlist(lapply(strsplit(sam_gene_gt$VARIANT, split = ":"), function(x) paste(x[3], x[4], sep = "_")))
          #  sam_gene_gt <- sam_gene_gt[which(vep_cod %in% ref_alt),]
          
        }
        else{
          ##compute cohort specific MAF
          maf_vec_cont <- sum(ifelse(is.na(as.numeric(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
          maf_vec_case <- sum(ifelse(!is.na(as.numeric(sam_gene_gt$SAMPLE)) | 
                                       grepl("^CR|^LK", as.character(sam_gene_gt$SAMPLE)), 1, 0))/(2*length(p_vec))
          #maf_vec <- (maf_vec_cont + maf_vec_case)/(2*(1572 + 1110))
          ##genotype matrix  
          sam_gene_gt$add_mod <- as.numeric(sam_gene_gt$GT)
          sam10 <- ifelse(Ex_samp_id %in% sam_gene_gt$SAMPLE, 1, 0)
          sam10[which(sam10 != 0)] <- sam_gene_gt$add_mod ##additive model
          ##account for discrepancy in vep_consequence
          #vep_con <- names(sort(table(unlist(strsplit(sam_gene_gt$vep_consequence, split = "&"))),decreasing=TRUE)[1])
          # sam_gene_gt$vep_consequence <- vep_con
          
          
          samp_vec[[m]] <- c(ftemp_tab_var_id[m],
                             unique(sam_gene_gt$gene_symbol), 
                             unique(sam_gene_gt$vep_consequence), 
                             unique(as.character(sam_gene_gt$auto_call)),
                             as.numeric(unique(sam_gene_gt$comb_score)), as.numeric(maf_vec_case),
                             as.numeric(maf_vec_cont), sam10)
        }
      }
      samp_vec_mat_uni <- do.call("rbind.data.frame", samp_vec)
      colnames(samp_vec_mat_uni) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                      "coh_MAF_cont", Ex_samp_id)
      if(exists("samp_vec_con")){
        samp_vec_mat_con <- do.call("rbind.data.frame", samp_vec_con)
        colnames(samp_vec_mat_con) <- c("VARIANT", "cpx_name", "vep_consequence", "auto_call", "comb_score", "coh_MAF_case",
                                        "coh_MAF_cont", Ex_samp_id)
        samp_vec_mat <- rbind.data.frame(samp_vec_mat_uni, samp_vec_mat_con)
      }else{
        samp_vec_mat <- samp_vec_mat_uni
      }
      print(dim(samp_vec_mat))
      ##Intracohort filter  : equivalent to ~ 5/1661*2 for case ; ~ 5/3209*2 for control
      ##changed to uniform intra_cohort_MAF filter
      ## change to entire cohort MAF filter to 3/(4849*2)##latest Aug.31
      #     samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_case)) <= 0.0015 | 
      #                                    as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
      samp_vec_mat <- samp_vec_mat[!(as.numeric(as.character(samp_vec_mat$coh_MAF_case)) >= 0.00035 | 
                                       as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) >= 0.00035),]
      #  samp_vec_mat <- samp_vec_mat[as.numeric(as.character(samp_vec_mat$coh_MAF_cont)) <= 0.001,]
      
      if(is.null(dim(samp_vec_mat)) | dim(samp_vec_mat)[1] == 0) {
       
        skat_pvals_sex_pc12 <- NULL
  
        next
      }
      ##compute maf SKAT takes care of the AFs internally by assigning higher weights to rare variants 
      
      else {
        gene_mat_comb <- as.numeric(as.character(samp_vec_mat[,5]))
        gene_mat <- as.matrix(samp_vec_mat[,-c(1:7)])
        class(gene_mat) <- "numeric"
        
        ##gender + totex + pc1:4
        skat_pvals_sex_pc12 <- SKAT_run(geno_mat = gene_mat, gene = cpx_name, x=p_Data_noCH[,c(55:56,19:22)], p_vec, cust_weight = gene_mat_comb, rho = 1)
        
      }
      
    } ##end of nested else loop
    return(skat_pvals_sex_pc12)
  }##end of gene loop ; i loop   
  


Shelterin <- c("POT1", "TINF2", "TERF1", "SMARCAL1", "STAG3")
Sarcoma_genes <- c("TP53", "NF1", "SDHA", "SDHB", "SDHD")
CEP_HAUS_core <- c("CEP63", "CEP72", "HAUS4", "HAUS5", "MZT1", "SSNA1")

Breast_cancer_genes = c("BRCA2", "BRCA1", "PALB2")

MPNST_pos <- c("NF1", "LZTR1", "SDHA", "SDHB", "SDHC", "SDHD")

cpx_list <- list(Shelterin, CEP_HAUS_core, Breast_cancer_genes, Sarcoma_genes, MPNST_pos)
names(cpx_list) <- c("Shelterin", "Centrosome", "BRCA_genes", "SARC_genes", "MPNST")

##Parallelised SKAT  
library(doParallel)
library(doMC)
registerDoMC(30)


res20_cpx_var_pc <- list()
#genes <- unique(fil_tab$gene_symbol)
#genes <- names(cpx_list)
system.time(res20_cpx_var_pc <- foreach(i=1:length(cpx_list), .errorhandling = 'remove') %dopar% 
{para_SKAT_cpx_var(cpx_list[[i]], names(cpx_list)[i], dbnsfp_cpx_samp_01)}) ##from POPMAX : line 454

res20_cpx_var_pc_df = do.call("rbind.data.frame", res20_cpx_var_pc)
res20_cpx_var_pc_df$p.adj = p.adjust(res20_cpx_var_pc_df$pval_SKATbin)
